openapi/src/main/java/io/helidon/openapi/Serializer.java

```diff
@@ -0,0 +1,349 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import io.smallrye.openapi.api.models.OpenAPIImpl;
+import io.smallrye.openapi.runtime.io.OpenApiSerializer;
+import org.eclipse.microprofile.openapi.models.Extensible;
+import org.eclipse.microprofile.openapi.models.OpenAPI;
+import org.eclipse.microprofile.openapi.models.parameters.Parameter;
+import org.yaml.snakeyaml.DumperOptions;
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.introspector.Property;
+import org.yaml.snakeyaml.nodes.MappingNode;
+import org.yaml.snakeyaml.nodes.Node;
+import org.yaml.snakeyaml.nodes.NodeId;
+import org.yaml.snakeyaml.nodes.NodeTuple;
+import org.yaml.snakeyaml.nodes.ScalarNode;
+import org.yaml.snakeyaml.nodes.Tag;
+import org.yaml.snakeyaml.representer.Representer;
+
+/**
+ * Expresses an existing {@code OpenAPI} instance as an OpenAPI document. This implementation uses
+ * SnakeYAML to write OpenAPI documents from the SmallRye MP OpenAPI model interfaces and classes
+ * while suppressing tags that would indicate the SmallRye classes -- we don't want to
+ * suggest that the output can only be read into the SmallRye implementation.
+ */
+class Serializer {
+
+    private static final DumperOptions YAML_DUMPER_OPTIONS = new DumperOptions();
+    private static final DumperOptions JSON_DUMPER_OPTIONS = new DumperOptions();
+
+    private static final Logger LOGGER = Logger.getLogger(Serializer.class.getName());
+
+    private Serializer() {
+    }
+
+    static {
+        YAML_DUMPER_OPTIONS.setIndent(2);
+        YAML_DUMPER_OPTIONS.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
+
+        JSON_DUMPER_OPTIONS.setDefaultFlowStyle(DumperOptions.FlowStyle.FLOW);
+        JSON_DUMPER_OPTIONS.setPrettyFlow(true);
+        JSON_DUMPER_OPTIONS.setDefaultScalarStyle(DumperOptions.ScalarStyle.DOUBLE_QUOTED);
+        JSON_DUMPER_OPTIONS.setSplitLines(false);
+    }
+
+    static void serialize(Map<Class<?>, ExpandedTypeDescription> types, Map<Class<?>, ExpandedTypeDescription> implsToTypes,
+            OpenAPI openAPI, OpenApiSerializer.Format fmt,
+            Writer writer) {
+        if (fmt == OpenApiSerializer.Format.JSON) {
+            serialize(types, implsToTypes, openAPI, writer, JSON_DUMPER_OPTIONS, DumperOptions.ScalarStyle.DOUBLE_QUOTED);
+        } else {
+            serialize(types, implsToTypes, openAPI, writer, YAML_DUMPER_OPTIONS, DumperOptions.ScalarStyle.PLAIN);
+        }
+    }
+
+    private static void serialize(Map<Class<?>, ExpandedTypeDescription> types,
+            Map<Class<?>, ExpandedTypeDescription> implsToTypes, OpenAPI openAPI, Writer writer,
+            DumperOptions dumperOptions,
+            DumperOptions.ScalarStyle stringStyle) {
+
+        Yaml yaml = new Yaml(new CustomRepresenter(types, implsToTypes, dumperOptions, stringStyle), dumperOptions);
```

Comment:
Hazy old memories of making sure that SnakeYAML's `SafeConstructor` or something is in play; I'll try to dig for more details.

Reply:
IIRC that's if you want to restrict the document-to-object mapping to objects that come with Java (i.e., not your own). (see https://bitbucket.org/asomov/snakeyaml/wiki/Documentation and search for SafeConstructor)